Issue #578 - Add tag support for godog tests into make integration

agajdosi opened the pull request:
To run integration tests on only specified features and scenarios, one can run e.g.: `make integration TEST_TAGS="openshift docker proxy"`.
When TEST_TAGS is not specified, then tests run with native tag: `basic`, which is bind to smoke test in basic.feature file. This means that when we add more time consuming feature files , one will be still able to run quick check with just `make integration`. Or run smoke test and feature close to specific issue he is dealing with. For example `make integration TEST_TAGS="smoke proxy"` when one is dev dealing with proxy stuff.

hferentschik commented on Makefile:
```diff
@@ -117,7 +118,7 @@ test: vendor
 .PHONY: integration
 integration: $(MINISHIFT_BINARY)
-	go test -timeout 3600s $(REPOPATH)/test/integration --tags=integration -v -args --binary $(MINISHIFT_BINARY)
+	go test -timeout 3600s $(REPOPATH)/test/integration --tags=integration -v -args --binary $(MINISHIFT_BINARY) --tags "$(TEST_TAGS)"
```
does that not force me to always specify a tag? Should there not be a way run all tests w/o specifying a tag? Trying to understand your thinking here.

hferentschik commented on test/integration/basic_test.go:
```diff
@@ -181,11 +179,19 @@ func FeatureContext(s *godog.Suite) {
+	flag.StringVar(&givenTags, "tags", "", "Tags of tests to pass to godog")
+	flag.Parse()
+
+	givenTags = strings.Replace(givenTags, " ", ", ", -1)
```
Again, should this not be a bit more defensive? What if no tag is given? Will it run all or nothing?

agajdosi replied:
@hferentschik If no tags are given `make integration TEST_TAGS=""` then all tests are executed. Currently only problem I have found is this case: `make integration TEST_TAGS="basic openshift "` with extra space or comma at the end it fails.

hferentschik replied:
A couple of things, I think this should be guarded with an if statement, if no tags are given the 'tag' argument should not even be set in the good.Options.

Secondly, why do we not pass the tags comma separated in the first place?

praveenkumar commented on test/integration/features/basic.feature:
```diff
@@ -8,7 +9,7 @@ Feature: Basic
      Then Minishift should have state "Running"
       And Minishift should have a valid IP address
-
+  
```
This added only because of a whitespace so might be good to remove that space.

agajdosi replied:
@praveenkumar yes, thank you for a catch

praveenkumar commented on Makefile:
```diff
@@ -25,6 +25,7 @@ ifeq ($(GOOS),windows)
 	IS_EXE := .exe
 endif
 MINISHIFT_BINARY ?= $(GOPATH)/bin/minishift$(IS_EXE)
+TEST_TAGS ?= basic
```
@hferentschik so looks like if user doesn't define a tag during run of integration then it only consider to run basic tests but I do anticipate it should be running all the test instead of basic when no tags are defined.

agajdosi replied:
@praveenkumar @hferentschik My idea here was to make it easy to use - having basic test as a smoke test which you can run while on coffee break etc. And to not force dev to type a lot: `make integration TEST_TAGS=basic`, I put basic test as native option there. However I agree with you that this is not so obvious setup. Let me know about this and I will update this PR to run all tests on default.

hferentschik replied:
So here is an idea, why don't we make this variable "CUCUMBER_OPTS" and make it per default the empty string. Then we pass the whole _CUCUMBER_OPTS_ to the _go test_ command. This means per default not tags are applied.

If you then want to run with tags, you do:

    $ make integration CUCUMBER_OPTS="--tags basic,smoke,whatever"

This might be also more generic in case we want to pass parameters.

agajdosi replied:
@hferentschik working on it now. Only I would change CUCUMBER_OPTS to GODOG_OPTS as except tags and feature paths, all other options are relevant only for godog, for example StopOnFailure, NoColors etc.

agajdosi commented:
@hferentschik Good point, I am going to add documentation on both new features of `make integration` in next update of this branch.

agajdosi commented on docs/source/developing/developing.adoc:
```diff
@@ -209,12 +209,40 @@ For more information about test options, run the `go test --help` command and re
+Please note that when multiple values are used for options in GODOG_OPTS, then those have to be separated with comma without a whitespace.
+While `-tags basic,openshift` will be parsed properly by make, from `-tags basic, openshift` only `@basic` tag will be used.  
+
+
```
Hi @Preeticp, I have tried to add documentation for new features I have added to `make integration` as proposed by @hferentschik. Can I ask you for a review of this? I have no experience with writing proper documentation so there are some mistakes I guess... Thankful for any suggestion. Just right now I have realized I have one extra line here.

hferentschik replied:
no worries. Looking good

Preeticp replied:
Sorry @agajdosi, just saw this, will review this.

hferentschik commented:
@agajdosi so you updated the pull request now, right? I guess we can take [WIP] off again.

agajdosi commented:
@hferentschik Yes, PR is updated. I have added docs into developing-integration as you suggested and created GODOG_OPTS argument in make, which supports most options available in godog Options. I dropped concurrency for which test suite is currently not prepared I guess.

hferentschik commented on Makefile:
```diff
@@ -28,6 +28,7 @@ ifeq ($(GOOS),windows)
 	IS_EXE := .exe
 endif
 MINISHIFT_BINARY ?= $(GOPATH)/bin/minishift$(IS_EXE)
+GODOG_OPTS ?= ""
```
:+1:

hferentschik commented on test/integration/basic_test.go:
```diff
@@ -181,11 +178,25 @@ func FeatureContext(s *godog.Suite) {
 }
 
 func TestMain(m *testing.M) {
+	flag.StringVar(&givenArgs, "minishift-args", "", "Arguments to pass to minishift")
```
Cool, now we can handle all potential flags. Personally I'd put the command line parsing into a separate function. But this will work.

hferentschik commented:
Merged, thanks.

@agajdosi, awesome work. Loving it. This is what I had in mind. I in particular like 'GODOG_OPTS="-format pretty"'.  I am tempted to make 'pretty' the default, but at the very least I would think it should be used on the CI servers, since one gets a better insight what's happening.

Also funny to see how much the setup "costs" and how fast the actual tests are. The bottle neck is the setup.
